packages/app/src/cli/services/dev/extension/bundler.ts

```diff
@@ -0,0 +1,84 @@
+import {ExtensionsPayloadStore} from './payload/store.js'
+import {ExtensionDevOptions} from '../extension.js'
+import {bundleExtension} from '../../extensions/bundle.js'
+import {abort, path, output} from '@shopify/cli-kit'
+import chokidar from 'chokidar'
+
+export interface WatchEvent {
+  path: string
+  type: 'build' | 'localization'
+}
+
+interface FileWatcherOptions {
+  devOptions: ExtensionDevOptions
+  payloadStore: ExtensionsPayloadStore
+}
+
+export interface FileWatcher {
+  close: () => void
+}
+
+export function setupBundlerAndFileWatcher(options: FileWatcherOptions) {
+  const abortController = new abort.Controller()
+
+  options.devOptions.extensions.forEach((extension) => {
+    const esbuild = bundleExtension({
+      minify: false,
+      outputBundlePath: extension.outputBundlePath,
+      sourceFilePath: extension.entrySourceFilePath,
+      environment: 'development',
+      env: options.devOptions.app.dotenv?.variables ?? {},
+      stderr: options.devOptions.stderr,
+      stdout: options.devOptions.stdout,
+      watchSignal: abortController.signal,
+      watch: (error, result) => {
+        output.debug(
+          `The Javascript bundle of the UI extension with ID ${extension.devUUID} has ${
+            error ? 'an error' : 'changed'
+          }`,
+        )
+
+        options.payloadStore
+          .updateExtension(extension, {
+            status: error ? 'error' : 'success',
+          })
+          .then((_) => {})
+          .catch((_) => {})
```

Comment:
Any specific reason why we call `then`, `catch` instead of using `try`/`catch`? Not sure if we follow this convention everywhere in the codebase, but I think we use `try`/`catch` blocks more. In this PR there are other occurrences of this if you search by `.catch`.

Reply:
@pepicrft answered this offline, `esbuild` doesn't allow us to pass `async` functions. It might still be worth adding some logs here and also when calling `then`.